test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ModelBuilder/ExplicitModelBuilderTests.cs

```diff
@@ -135,56 +144,60 @@ private static IEdmModel GetExplicitEdmModel()
+              {
+                  object id;
+                  entityContext.EdmObject.TryGetPropertyValue("ID", out id);
+                  return new Uri(entityContext.GetUrlHelper().Link(ODataTestConstants.DefaultRouteName,
+              new
+              {
+                  odataPath = ResourceContextHelper.CreateODataLink(entityContext,
+                      new EntitySetSegment(entityContext.NavigationSource as IEdmEntitySet),
+                      new KeySegment(new[] { new KeyValuePair<string, object>("ID", id) }, entityContext.StructuredType as IEdmEntityType, null),
+                      new NavigationPropertySegment(navigationProperty, null))
+              }));
+              }, true);
 
             suppliers.HasNavigationPropertiesLink(
-                supplier.NavigationProperties,
-                (entityContext, navigationProperty) =>
-                {
-                    object id;
-                    entityContext.EdmObject.TryGetPropertyValue("ID", out id);
-                    return new Uri(entityContext.GetUrlHelper().Link(
-                ODataTestConstants.DefaultRouteName,
-                new
-                {
-                    odataPath = ResourceContextHelper.CreateODataLink(entityContext,
-                        new EntitySetSegment(entityContext.NavigationSource as IEdmEntitySet),
-                        new KeySegment(new[] { new KeyValuePair<string, object>("ID", id) }, entityContext.StructuredType as IEdmEntityType, null),
-                        new NavigationPropertySegment(navigationProperty, null))
-                }));
-                }, true);
+              supplier.NavigationProperties,
+              (entityContext, navigationProperty) =>
+              {
+                  object id;
+                  entityContext.EdmObject.TryGetPropertyValue("ID", out id);
+                  return new Uri(entityContext.GetUrlHelper().Link(
+              ODataTestConstants.DefaultRouteName,
+              new
+              {
+                  odataPath = ResourceContextHelper.CreateODataLink(entityContext,
+                      new EntitySetSegment(entityContext.NavigationSource as IEdmEntitySet),
+                      new KeySegment(new[] { new KeyValuePair<string, object>("ID", id) }, entityContext.StructuredType as IEdmEntityType, null),
+                      new NavigationPropertySegment(navigationProperty, null))
+              }));
+              }, true);
+
+            var function = supplier.Function("GetAddress").Returns<Address>().AddDerivedTypeConstraint<Address>();
+            function.Parameter<int>("value");
+
+            var action = modelBuilder.Action("GetAddress").Returns<Address>().AddDerivedTypeConstraint<Address>();
+            action.Parameter<Supplier>("supplier").AddDerivedTypeConstraint<ToiletPaperSupplier>();
 
             return modelBuilder.GetEdmModel();
         }
 
         [Fact]
-
         public async Task VerifyMetaDataIsGeneratedCorrectly()
```

Comment:
>VerifyMetaDataIsGeneratedCorrectly() [](start = 26, length = 36)

Add test cases in "MetadataControllerTest.cs" to verify the annotation showing in the xml

Reply:
need E2E test cases to verify:
1) Uri parsing
2) Serialization

---
In reply to: [358425347](https://github.com/OData/WebApi/pull/1977#discussion_r358425347) [](ancestors = 358425347)